Added floco lock and cs8recovery procedure

vshekar opened the pull request:
Added commands for flocos to stop and restart govmon and LSDC GUI

When the robot controller (CS8) needs to be rebooted, several items need to be reset so that the robot is "aware" of it's environment. To do this we reset the following:
Governor
Gripper sensors
Dewar position
Dewar ports

RobertSchaffer1 commented on daq_macros.py:
```diff
@@ -259,6 +259,24 @@ def run_robot_recovery_procedure():
   # Park Gripper and cool gripper
   robot_lib.cooldownGripper()
 
+def recoverCS8():
```
Does this get called anywhere?

RobertSchaffer1 replied:
Or does it just get called by operators?

vshekar replied:
Just gets called by flocos yes

RobertSchaffer1 commented on daq_macros.py:
```diff
@@ -259,6 +259,24 @@ def run_robot_recovery_procedure():
+  logger.info("Setting robot state to SE")
+  gov_lib.setGovRobot(gov_robot, 'SE')
+  logger.info("Setting robot state to M")
+  gov_lib.setGovRobot(gov_robot, 'M')
```
Why are we going SE->M->SE?

vshekar replied:
Edwin has details but I think M -> SE resets the robot arm somehow

RobertSchaffer1 replied:
Should just eliminate the first SE?

RobertSchaffer1 commented:
Is this ready for re-review? I saw the one DONOTMERGE in there so I was assuming it was a WIP, but let me know if it's otherwise.

vshekar commented:
> Is this ready for re-review? I saw the one DONOTMERGE in there so I was assuming it was a WIP, but let me know if it's otherwise.

Hey Rob, sorry it's not ready for re-review. I have to clean up the commits and include a few of the changes I made last week with Edwin. I'll ping you when its ready. Most likely this week

vshekar commented:
Additions to the PR:
- Moved the cryostream device to its own file
- Added virtual dewar device tracks rotation as well as puck sensors to detect puck lifts

RobertSchaffer1 commented on daq_lib.py:
```diff
@@ -384,18 +424,44 @@ def waitBeam():
+    currentRequest = db_lib.popNextRequest(daq_utils.beamline)
+    if (getBlConfig("queueCollect") == 1): 
+      if (getBlConfig(BEAM_CHECK) == 1):
+        waitBeam()
```
Double-checking, BEAM_CHECK == 1 means you need to wait?

RobertSchaffer1 commented on daq_macros.py:
```diff
@@ -259,11 +259,87 @@ def run_robot_recovery_procedure():
   """ Generic recovery procedure to be used during automated
   collection"""
   # Recover robot
+  logging.info("Running robot recovery procedure (nuclear option)")
```
logger vs. logging?

vshekar replied:
Good catch, fixed in https://github.com/NSLS2/lsdc/pull/399/commits/016572acc673dfedca454f166725235945b69d11

RobertSchaffer1 commented on mxbluesky/devices/generic.py:
```diff
@@ -1,5 +1,7 @@
-from ophyd import Device, EpicsMotor, EpicsSignal, EpicsSignalRO, PVPositionerIsClose
+from ophyd import Device, EpicsMotor, EpicsSignal, EpicsSignalRO
+from ophyd.status import SubscriptionStatus
+from ophyd import DynamicDeviceComponent as DDC
```
FYI you can do this on the same line as the other ophyd imports, not sure which black prefers though.

vshekar replied:
Makes sense to import together, fixed in https://github.com/NSLS2/lsdc/pull/399/commits/016572acc673dfedca454f166725235945b69d11

RobertSchaffer1 commented on mxbluesky/devices/generic.py:
```diff
@@ -36,17 +38,64 @@ def __init__(self, *args, **kwargs):
+
+    def rotate(self, rotation_angle, absolute=True):
+        def check_value_sink(*, old_value, value, **kwargs):
+            "Return True when the movement is complete, False otherwise."
```
I think these messages are backwards?

vshekar replied:
I think I copied code you had provided 😄 
Should I switch it? It seems to be working as is. It's at the beamlines

RobertSchaffer1 replied:
Yeah, it's confusing, just swap the comments between the two and I think it's fine.

RobertSchaffer1 commented on start_bs.py:
```diff
@@ -141,31 +144,43 @@ class SampleXYZ(Device):
+    from mxtools.raster_flyer import MXRasterFlyer
+    from embl_robot import EMBLRobot
+    import setenergy_lsdc
+
```
Should we merge some of these with the above? Maybe one above both with shared imports and if beamline in ["fmx", "amx"].

vshekar replied:
That's a good idea. I'll merge them in a future PR
